### PR TITLE
Add appender support to autodiscover

### PR DESCRIPTION
**Pull request by vjsamuel**
Appenders have the ability to append additional info into the configuration. There are two appenders that are part of this PR:

`config` appender: This appender can append a templatized config into all of the generated configs by either templates or builders and can be used by any Beat.

Sample config:
```
appenders:
  - type: config
     - condition.equals:
          kubernetes.labels.app: "prometheus"
        config: 
           fields:
             test: success  
```
`kubernetes.token` appender: There are endpoints in kubernetes that require bearer token authentication. This appender can add `headers.Authorization` to a metricset configuration. This can be used to collect metrics from Kube API server.

```
appenders:
  - type: "kubernetes.token"
```

### CODE REVIEW

**Comment by ruflin** [comment]
jenkins, test it

**Comment by exekias** [comment]
I'm wondering if this construct should be moved to the module level instead, for a wider usage. At the end, you could need this exact functionality outside autodiscover.

WDYT?

**Comment by vjsamuel** [comment]
I thought of the same as well. Any specific thoughts on how it should look?

**Comment by ruflin** [comment]
I definitively think the features we brought to k8s so far can also be useful in a more generic way like mentioned above. I'm wondering if we could go forward with the above implementation for now, mark it as beta, learn from it and then extract it to be also reused in other parts?

**Comment by exekias** [comment]
The thing is this specific appender would be cool also for some commonly static Metricbeat modules, like kubernetes, prometheus or even http, as it gives them auth inside the cluster.

I could also implement that at the http helper level, and expose the settings only on these modules, but appender interface sounds like the right solution to the problem

**Comment by ruflin** [comment]
@exekias WDYT about moving forward. Getting it in as is and abstract it later or you think that will be tricky?

**Comment by vjsamuel** [comment]
@ruflin  i have added experimental tag to all the builders/providers and appenders.

File: libbeat/autodiscover/appenders/config/config.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	// Config key doesnt have an array of config objects
+	if !ok {
+		return
+	}
+	for _, configMap := range c.configMaps {
+		if configMap.condition == nil || configMap.condition.Check(common.MapStr(event)) == true {
+			// Merge the template with all the configs
+			for _, cfg := range cfgs {
+				cf := common.MapStr{}
+				err := cfg.Unpack(&cf)
```
Comment (exekias) [inline libbeat/autodiscover/appenders/config/config.go]: This error is not being checked

**Comment by exekias** [comment]
Ok, let's get this in, we can make it general later

I left just a minor comment

**Comment by vjsamuel** [comment]
@exekias @ruflin done with all review comments.

File: libbeat/autodiscover/appenders/config/config.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+				if err != nil {
+					logp.Debug("config", "unable to unpack config due to error: %v", err)
```
[the remark this answers is not in the document]
Comment (vjsamuel) [reply libbeat/autodiscover/appenders/config/config.go]: this can be left as is.

**Comment by vjsamuel** [comment]
@andrewkroh i have incorporated review comments.

**Comment by andrewkroh** [comment]
Thanks.

Looks like we still need documentation on Appenders.

And since there is config associated with appenders we should have it in the [reference config](https://github.com/elastic/beats/blob/master/libbeat/_meta/config.reference.yml). @exekias How come `add_kubernetes_metadata` isn't in the reference config?

**Comment by vjsamuel** [comment]
@andrewkroh we talked about the documentation and some consensus needs to be arrived at for the structuring of autodiscover documentation as there are providers, templates, builders and appenders that needs documentation at the moment. i would reserve that for a separate PR.

**Comment by andrewkroh** [comment]
I think we could at least document the basics of what it does, when/why to use it, and how to configure it without knowing what the exact structure final structure will be by adding a new "Appenders" section to the [autodiscover](https://github.com/elastic/beats/pull/6469#issuecomment-371566480) topic. If we produce the content our talented technical writers can later shape it and give it more structure.

Basically I'm dislike adding new features without docs.

**Comment by vjsamuel** [comment]
Sure thing. Let me add a documentation commit.

**Comment by andrewkroh** [comment]
jenkins, test it
